exporter/loadbalancingexporter/consistent_hashing.go

```diff
@@ -0,0 +1,168 @@
+type position uint32
+
+// ringItem connects a specific angle in the ring with a specific endpoint.
+type ringItem struct {
+	pos      position
+	endpoint string
+}
+
+// hashRing is a consistent hash ring following Karger et al.
+type hashRing struct {
+	// ringItems holds all the positions, used for the lookup the position for the closest next ring item
+	items []ringItem
+}
+
+// newHashRing builds a new immutable consistent hash ring based on the given endpoints.
+func newHashRing(endpoints []string) *hashRing {
+	items := positionsForEndpoints(endpoints, defaultWeight)
+	return &hashRing{
+		items: items,
+	}
+}
+
+// endpointFor calculates which backend is responsible for the given traceID
+func (h *hashRing) endpointFor(traceID pdata.TraceID) string {
+	b := traceID.Bytes()
+	hasher := crc32.NewIEEE()
+	hasher.Write(b[:])
+	hash := hasher.Sum32()
+	pos := hash % maxPositions
+
+	return h.findEndpoint(position(pos))
+}
+
+// findEndpoint returns the "next" endpoint starting from the given position
+func (h *hashRing) findEndpoint(pos position) string {
+	ringSize := len(h.items)
+	left, right := h.items[:ringSize/2], h.items[ringSize/2:]
+	found := bsearch(pos, left, right)
+	return found.endpoint
+}
+
+// bsearch is a binary search-like algorithm, returning the closest "next" item instead of an exact match
+func bsearch(pos position, left []ringItem, right []ringItem) ringItem {
+	// if it's the last item of the left side, return it
+	if left[len(left)-1].pos == pos {
+		return left[len(left)-1]
+	}
+
+	// if it's the first item of the right side, return it
+	if right[0].pos == pos {
+		return right[0]
+	}
+
+	// if we want a higher angle than the highest from the ring, the first angle is the right one
+	if pos > right[len(right)-1].pos {
+		return left[0]
+	}
+
+	// if the requested position is higher than the highest in the left, the item is in the right side
+	if pos > left[len(left)-1].pos {
+		size := len(right)
+		if size == 1 {
+			return right[0]
+		}
+
+		l, r := right[:size/2], right[size/2:]
+		return bsearch(pos, l, r)
+	}
+
+	// not on the right side, has to be on the left side
+	size := len(left)
+	if size == 1 {
+		return left[0]
+	}
+	l, r := left[:size/2], left[size/2:]
+	return bsearch(pos, l, r)
+}
+
+// positionFor calculates all the positions in the ring based. The numPoints indicates how many positions to calculate.
+// The slice length of the result matches the numPoints.
+func positionsFor(endpoint string, numPoints int) []position {
+	res := []position{}
+	for i := 0; i < numPoints; i++ {
+		h := crc32.NewIEEE()
+		h.Write([]byte(endpoint))
+		h.Write([]byte{byte(i)})
+		hash := h.Sum32()
+		pos := hash % maxPositions
+		res = append(res, position(pos))
+	}
+
+	return res
+}
+
+// positionsForEndpoints calculates all the positions for all the given endpoints
+func positionsForEndpoints(endpoints []string, weight int) []ringItem {
+	items := []ringItem{}
```

Comment:
same, you either don't allocate the items at all (not just an empty slice), or allocate the size if you know it.

Reply:
Done.